src/doc-utils.js

```diff
@@ -41,26 +51,92 @@ function isEmpty(n) {
 }
 
 function getFirstString(doc) {
-  return iterDoc(doc, (type, doc) => {
-    if (type === "string" && doc.trim().length !== 0) {
-      return doc;
+  let firstString = null;
+  traverseDoc(doc, doc => {
+    if (
+      typeof doc === "string" && doc.trim().length !== 0 && firstString === null
+    ) {
+      firstString = doc;
+    }
+  });
+  return firstString;
+}
+
+function isLineNext(doc) {
+  let flag = null;
+  traverseDoc(doc, doc => {
+    if(flag === null) {
+      if (typeof doc === "string") {
+        flag = false;
+      }
+      if (doc.type === "line") {
+        flag = true;
+      }
     }
   });
+  return !!flag;
 }
 
-function hasHardLine(doc) {
-  // TODO: If we hit a group, check if it's already marked as a
-  // multiline group because they should be marked bottom-up.
-  return !!iterDoc(doc, (type, doc) => {
-    switch (type) {
+function willBreak(doc) {
+  let willBreak = false;
+  traverseDoc(doc, doc => {
+    switch (doc.type) {
+      case "group":
+        if(doc.break) {
+          willBreak = true;
+        }
```

Comment:
> I'm not sure it's necessary, since it complicates traverseDoc a lot.

It's actually not very invasive:

```js
function traverseDoc(doc, fn) {
  var isStopped = false;

  function traverseDocRec(doc) {
    if (isStopped) {
      return;
    }

    if (doc.type === 'x') {
      isStopped = isStopped || fn(doc);
      traverseDocRec(...);
    }
  }

  traverseDocRec(doc);
}
```